src/main/java/net/satago/tapestry5/jpa/TransactionalUnit.java

```diff
@@ -33,7 +33,7 @@
 
```

Comment:
We've been discussing this fix offline with @pedrocborges2 long time ago, but never returned to it after that.

The problem we're trying to solve here is that if you try to create a new `transactionalUnit` from `afterCommit` callback there will still be previous (already committed) top level `transactionalUnit` in the `currentUnit` stack. And as a result of that this newly created unit will never be committed.

So instead of playing with the state of current transaction as in `transaction.isActive()` (which, by the way, *may* be changed outside of transactionalUnits API via JPA interfaces) we could do `currentUnit.get().pop()` before invoking callbacks.

This should work fine for `afterCallback`, but the question is still open for what we should do if user is trying to run another transactional unit from `beforeCommit` callback. I don't have any good answer for this right now. This is an edge case and we don't do this in our applications, so we could either document this better somewhere and/or forbid creating new nested units from the callback(s) by throwing an exception.

What do you think?